nifi-nar-bundles/nifi-standard-services/nifi-lookup-services-bundle/nifi-lookup-services/src/main/java/org/apache/nifi/lookup/DistributedMapCacheRecordLookupService.java

```diff
@@ -0,0 +1,183 @@
+        return new PropertyDescriptor.Builder().name(propertyDescriptorName).required(false).dynamic(true).addValidator(Validator.VALID)
+                .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES).build();
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        cache = context.getProperty(PROP_DISTRIBUTED_CACHE_SERVICE).asControllerService(DistributedMapCacheClient.class);
+        charset = Charset.forName(context.getProperty(CHARACTER_ENCODING).getValue());
+        readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(PROP_DISTRIBUTED_CACHE_SERVICE);
+        descriptors.add(CHARACTER_ENCODING);
+        descriptors.add(RECORD_READER);
+        return descriptors;
+    }
+
+    @Override
+    public Set<String> getRequiredKeys() {
+        return REQUIRED_KEYS;
+    }
+
+    public static class StringSerializer implements Serializer<String> {
+        @Override
+        public void serialize(final String value, final OutputStream out) throws SerializationException, IOException {
+            out.write(value.getBytes(charset));
+        }
+    }
+
+    public static class CacheValueDeserializer implements Deserializer<byte[]> {
+
+        @Override
+        public byte[] deserialize(final byte[] input) throws DeserializationException, IOException {
+            if (input == null || input.length == 0) {
+                return null;
+            }
+            return input;
+        }
+    }
+
+    @Override
+    public Optional<Record> lookup(Map<String, Object> coordinates) throws LookupFailureException {
+        final Optional<String> docId = Optional.ofNullable(coordinates.get(KEY)).map(Object::toString);
+        if (!docId.isPresent()) {
+            return Optional.empty();
+        }
+        final Optional<InputStream> inputStream;
+        try {
+            Optional<byte[]> byteArrayOutput = Optional.ofNullable(cache.get(docId.get(), keySerializer, valueDeserializer));
+            if (byteArrayOutput.isPresent()) {
+                inputStream = Optional.ofNullable(new ByteArrayInputStream(cache.get(docId.get(), keySerializer, valueDeserializer)));
+            } else {
+                return Optional.empty();
+            }
+        } catch (IOException ie) {
+            throw new LookupFailureException("Failed to lookup from cachelookup IOException using this coordinates: " + coordinates);
+        } catch (Exception e) {
+            throw new LookupFailureException("Failed to lookup from cachelookup using this coordinates: " + coordinates);
+        }
+
+        final Optional<Tuple<Exception, RecordReader>> errOrReader = inputStream.map(in -> {
```

Comment:
I get what you're trying to do here, but I think the use of Optionals here is making the flow more complex than what it needs to be. Based on the prior code blocks, `inputStream` should be non-null already. I think refactoring this to not use Optionals and Tuples at this point would make it a lot simpler.